src/core/portal/portal.ts

```diff
@@ -0,0 +1,194 @@
+   * This is different from where the component *renders*, which is determined by the PortalHost.
+   * The origin necessary when the host is outside of the Angular application context.
+   */
+  public origin: ElementRef;
+
+  constructor(component: Type, origin: ElementRef = null) {
+    super();
+    this.component = component;
+    this.origin = origin;
+  }
+}
+
+
+/**
+ * A `TemplatePortal` is a portal that represents some embedded template (TemplateRef).
+ */
+export class TemplatePortal extends Portal<Map<string, any>> {
+  /** The embedded template that will be used to instantiate an embedded View in the host. */
+  templateRef: TemplateRef;
+
+  /** Additional locals for the instantiated embedded view. */
+  locals: Map<string, any> = new Map<string, any>();
+
+  constructor(template: TemplateRef) {
+    super();
+    this.templateRef = template;
+  }
+
+  get origin(): ElementRef {
+    return this.templateRef.elementRef;
+  }
+
+  attach(host: PortalHost, locals?: Map<string, any>): Promise<Map<string, any>> {
+    this.locals = locals == null ? new Map<string, any>() : locals;
+    return super.attach(host);
+  }
+
+  detach(): Promise<void> {
+    this.locals = new Map<string, any>();
+    return super.detach();
+  }
+}
+
+
+/**
+ * A `PortalHost` is an space that can contain a single `Portal`.
+ */
+export interface PortalHost {
+  attach(portal: Portal<any>): Promise<any>;
+
+  detach(): Promise<any>;
+
+  dispose(): void;
+
+  hasAttached(): boolean;
+}
+
+
+/**
+ * Partial implementation of PortalHost that only deals with attaching either a
+ * ComponentPortal or a TemplatePortal.
+ */
+export abstract class BasePortalHost implements PortalHost {
+  /** The portal currently attached to the host. */
+  private attachedPortal_: Portal<any>;
+
+  /** A function that will permanently dispose this host. */
+  private disposeFn_: () => void;
+
+  /** Whether this host has already been permanently disposed. */
+  private isDisposed_: boolean = false;
+
+  /** Whether this host has an attached portal. */
+  hasAttached() {
+    return this.attachedPortal_ != null;
+  }
+
+  attach(portal: Portal<any>): Promise<any> {
+    if (portal == null) {
+      throw new BaseException('Must provide a portal to attach');
+    }
+
+    if (this.hasAttached()) {
+      throw new BaseException('A portal is already attached');
+    }
+
+    if (this.isDisposed_) {
+      throw new BaseException('This PortalHost has already been disposed');
+    }
+
+    if (portal instanceof ComponentPortal) {
+      this.attachedPortal_ = portal;
+      return this.attachComponentPortal(portal);
+    } else if (portal instanceof TemplatePortal) {
+      this.attachedPortal_ = portal;
+      return this.attachTemplatePortal(portal);
+    }
+
+    throw new BaseException(
+        'Attempting to attach an unknown Portal type. ' +
+        'BasePortalHost accepts either a ComponentPortal or a TemplatePortal.');
```

Comment:
Maybe we could use a template string here for multi-line rather than concatenation?